<a href="https://colab.research.google.com/github/kKawsarAlam/Feature-Engineering/blob/main/Label_Encoding.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Label Encoding**   

Label encoding is a data preprocessing technique in machine learning that converts categorical text data into numerical integers, assigning a unique number (e.g., 0, 1, 2) to each category. It enables algorithms to process non-numeric data, such as converting "Red", "Blue", "Green" into \(0,1,2\), but assumes an ordered relationship between categories, which may not always exist.  

**Key Aspects of Label Encoding:**   

Purpose: Converts categorical labels into a numeric form for model input.   
Methodology: Assigns a unique integer to each unique category in a dataset.  
Use Case: Often used for target variables (y) in classification problems.   
Limitation: It implies an ordinal relationship (e.g.,2>1>0), which can mislead algorithms if the categories are nominal (e.g., country names, colors).  
Example: In a feature with ['Low', 'Medium', 'High'], label encoding might assign (0,1,2) to represent the order.

In [96]:
# Import Libraries
import pandas as pd
import numpy as np

print("Successfully Libraries Imported")

Successfully Libraries Imported


In [110]:
# Load Dataset
df = pd.read_csv('/content/home_prices_simple.csv')
df

town  area   price
0   monroe township  2600  550000
1   monroe township  3000  565000
2   monroe township  3200  610000
3   monroe township  3600  680000
4   monroe township  4000  725000
5      west windsor  2600  585000
6      west windsor  2800  615000
7      west windsor  3300  650000
8      west windsor  3600  710000
9      robbinsville  2600  575000
10     robbinsville  2900  600000
11     robbinsville  3100  620000
12     robbinsville  3600  695000

**Label Encoding**  

**Method 1: Using map()**

In [98]:
label_map = {
    'monroe township': 0,
    'west windsor': 1,
    'robbinsville': 2
}

df['town_encoded'] = df['town'].map(label_map)
df


town  area   price  town_encoded
0   monroe township  2600  550000             0
1   monroe township  3000  565000             0
2   monroe township  3200  610000             0
3   monroe township  3600  680000             0
4   monroe township  4000  725000             0
5      west windsor  2600  585000             1
6      west windsor  2800  615000             1
7      west windsor  3300  650000             1
8      west windsor  3600  710000             1
9      robbinsville  2600  575000             2
10     robbinsville  2900  600000             2
11     robbinsville  3100  620000             2
12     robbinsville  3600  695000             2

**Method 2: Using replace()**

In [28]:
df['town_encoded'] = df['town'].replace({
    'monroe township': 0,
    'west windsor': 1,
    'robbinsville': 2
})
df

/tmp/ipython-input-1690058813.py:1: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df['town_encoded'] = df['town'].replace({


town  area   price  town_encoded
0   monroe township  2600  550000             1
1   monroe township  3000  565000             1
2   monroe township  3200  610000             1
3   monroe township  3600  680000             1
4   monroe township  4000  725000             1
5      west windsor  2600  585000             0
6      west windsor  2800  615000             0
7      west windsor  3300  650000             0
8      west windsor  3600  710000             0
9      robbinsville  2600  575000             2
10     robbinsville  2900  600000             2
11     robbinsville  3100  620000             2
12     robbinsville  3600  695000             2

**Method 3: Auto-create labels**

In [111]:
# Get unique categories
categories = df['town'].unique()

In [112]:
# Create mapping automatically
label_map = {cat: idx for idx, cat in enumerate(categories)}

In [114]:
# Apply encoding
df['town_encoded'] = df['town'].map(label_map)
df

town  area   price  town_encoded
0   monroe township  2600  550000             0
1   monroe township  3000  565000             0
2   monroe township  3200  610000             0
3   monroe township  3600  680000             0
4   monroe township  4000  725000             0
5      west windsor  2600  585000             1
6      west windsor  2800  615000             1
7      west windsor  3300  650000             1
8      west windsor  3600  710000             1
9      robbinsville  2600  575000             2
10     robbinsville  2900  600000             2
11     robbinsville  3100  620000             2
12     robbinsville  3600  695000             2

**Using Sklearn LabelEncoder()**

In [77]:
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()

In [78]:
df['town_encoded'] = le.fit_transform(df['town'])
df


town  area   price  town_encoded
0   monroe township  2600  550000             0
1   monroe township  3000  565000             0
2   monroe township  3200  610000             0
3   monroe township  3600  680000             0
4   monroe township  4000  725000             0
5      west windsor  2600  585000             2
6      west windsor  2800  615000             2
7      west windsor  3300  650000             2
8      west windsor  3600  710000             2
9      robbinsville  2600  575000             1
10     robbinsville  2900  600000             1
11     robbinsville  3100  620000             1
12     robbinsville  3600  695000             1

In [99]:
# Drop some redundant columns so that it can not mix up
df = df.drop(columns=['town'])
df

area   price  town_encoded
0   2600  550000             0
1   3000  565000             0
2   3200  610000             0
3   3600  680000             0
4   4000  725000             0
5   2600  585000             1
6   2800  615000             1
7   3300  650000             1
8   3600  710000             1
9   2600  575000             2
10  2900  600000             2
11  3100  620000             2
12  3600  695000             2

**Model Traning**

In [100]:
from sklearn.linear_model import LinearRegression
model = LinearRegression()

In [101]:
X = df.drop('price', axis=1)
y = df['price']

In [102]:
Y = df.price
Y

0     550000
1     565000
2     610000
3     680000
4     725000
5     585000
6     615000
7     650000
8     710000
9     575000
10    600000
11    620000
12    695000
Name: price, dtype: int64

In [103]:
model.fit(X, Y)

LinearRegression()

In [104]:
# Predict the price for monroe township [0]
print("monroe township:", model.predict([[3400, 0]]))

monroe township: [648306.95795851]


/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


In [105]:
# Predict the price for west windsor [0]
print("west windsor:", model.predict([[2800, 2]]))

west windsor: [600774.00403892]


/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


In [106]:
# Predict the price for monroe township [0, 0]
print("robbinsville:", model.predict([[3000, 1]]))

robbinsville: [612138.05764641]


/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


In [107]:
# Model Accuracy
score = model.score(X, Y)
print(f"Model Score: {score:.2f}")

Model Score: 0.91


I applied Label Encoding (map) to my categorical features and trained the linear regression model. After evaluation, the model achieved an accuracy of 0.91.   

Better approach for Linear Regression:  


*   Use One-Hot Encoding for nominal categorical variables.
*   Linear regression can assign separate coefficients to each category, avoiding misleading order assumptions.


